code/datums/elements/tool_renaming.dm

```diff
@@ -0,0 +1,68 @@
+/**
+ * Renaming tool element
+ *
+ * When using this tool on an object with UNIQUE_RENAME,
+ * lets the user rename/redesc it.
+ */
+/datum/element/tool_renaming
+
+/datum/element/tool_renaming/Attach(datum/target)
+	. = ..()
+	if(!isitem(target))
+		return ELEMENT_INCOMPATIBLE
+
+	RegisterSignal(target, COMSIG_ITEM_INTERACTING_WITH_ATOM, PROC_REF(attempt_rename))
+
+/datum/element/tool_renaming/Detach(datum/source)
+	. = ..()
+	UnregisterSignal(source, COMSIG_ITEM_INTERACTING_WITH_ATOM)
+
+/datum/element/tool_renaming/proc/attempt_rename(datum/source, mob/living/user, atom/interacting_with, list/modifiers)
+	if(!isobj(interacting_with))
+		return NONE
+	
+	var/obj/renamed_obj = interacting_with
+
+	if(!(renamed_obj.obj_flags & UNIQUE_RENAME))
+		return NONE
+
+	INVOKE_ASYNC(src, PROC_REF(async_rename), user, renamed_obj)
+	return ITEM_INTERACT_SUCCESS
+
+/datum/element/tool_renaming/proc/async_rename(mob/living/user, obj/renamed_obj)
+	var/pen_choice = tgui_input_list(user, "What would you like to edit?", "Pen Setting", list(OPTION_RENAME, OPTION_DESCRIPTION, OPTION_RESET))
+	if(QDELETED(renamed_obj) || !user.can_perform_action(renamed_obj) || isnull(pen_choice))
+		return
+
+	switch(pen_choice)
+		if(OPTION_RENAME)
+			var/input = tgui_input_text(user, "What do you want to name [renamed_obj]?", "Object Name", "[renamed_obj.name]", MAX_NAME_LEN)
+			var/old_name = renamed_obj.name
+			if(QDELETED(renamed_obj) || !user.can_perform_action(renamed_obj))
+				return
+			if(input == old_name || !input)
+				to_chat(user, span_notice("You changed [renamed_obj] to... well... [renamed_obj]."))
+				return
+			renamed_obj.AddComponent(/datum/component/rename, input, renamed_obj.desc)
+			to_chat(user, span_notice("You have successfully renamed \the [old_name] to [renamed_obj]."))
+			ADD_TRAIT(renamed_obj, TRAIT_WAS_RENAMED, PEN_LABEL_TRAIT)
+			renamed_obj.update_appearance(UPDATE_NAME)
+
+		if(OPTION_DESCRIPTION)
+			var/input = tgui_input_text(user, "Describe [renamed_obj]", "Description", "[renamed_obj.desc]", 280)
+			var/old_desc = renamed_obj.desc
```

Comment:
```suggestion
			var/old_desc = renamed_obj.desc
			var/input = tgui_input_text(user, "Describe [renamed_obj]", "Description", old_desc, 280)
```

also `280` should really be a define-constant (maybe named `MAX_DESC_LEN`) in the same spot as `MAX_NAME_LEN` so it can be easily changed/re-used in numerous places